File: BackEnd/PHP/laravelshop/app/Modules/Shortform/Services/ShortformService.php
<?php

namespace App\Modules\Shortform\Services;

use App\Common\Exceptions\BusinessException;
use App\Models\Product;
use App\Models\Shortform;
use App\Models\ShortformComment;
use App\Models\ShortformLike;
use App\Models\ShortformProduct;
use App\Models\User;
use Symfony\Component\HttpFoundation\Response;

class ShortformService
{
    public function create(User $user, array $payload): array
    {
        $shortform = Shortform::query()->create([
            'user_id' => $user->id,
            'title' => $payload['title'],
            'video_url' => $payload['videoUrl'],
            'thumbnail_url' => $payload['thumbnailUrl'] ?? null,
            'transcode_status' => 'COMPLETED',
        ]);

        foreach ($payload['productIds'] ?? [] as $productId) {
            if (Product::query()->where('id', $productId)->exists()) {
                ShortformProduct::query()->create([
                    'shortform_id' => $shortform->id,
                    'product_id' => $productId,
                ]);
            }
        }

        return $this->serializeShortform($shortform->fresh());
    }

    public function list(?string $cursor, int $limit): array
    {
        $query = Shortform::query()->orderByDesc('id');
        if ($cursor !== null) {
            $query->where('id', '<', (int) $cursor);
        }

        $items = $query->limit($limit)->get();

        return [
            'items' => $items->map(fn (Shortform $shortform): array => $this->serializeShortform($shortform))->all(),
            'nextCursor' => $items->count() === $limit ? (string) $items->last()->id : null,
        ];
    }

    public function show(int $id): array
    {
        $shortform = $this->findShortform($id);
        $shortform->increment('view_count');

        return $this->serializeShortform($shortform->fresh());
    }

    public function toggleLike(User $user, int $id): array
    {
        $shortform = $this->findShortform($id);
        $like = ShortformLike::query()->where('shortform_id', $id)->where('user_id', $user->id)->first();

        if ($like) {
            $like->delete();
            $liked = false;
        } else {
            ShortformLike::query()->create(['shortform_id' => $id, 'user_id' => $user->id]);
            $liked = true;
        }

        $shortform->forceFill([
            'like_count' => ShortformLike::query()->where('shortform_id', $id)->count(),
        ])->save();

        return [
            'liked' => $liked,
            'likeCount' => $shortform->fresh()->like_count,
        ];
    }

    public function addComment(User $user, int $id, array $payload): array
    {
        $shortform = $this->findShortform($id);
        $comment = ShortformComment::query()->create([
            'shortform_id' => $shortform->id,
            'user_id' => $user->id,
            'content' => $payload['content'],
        ]);

        $shortform->forceFill([
            'comment_count' => ShortformComment::query()->where('shortform_id', $id)->count(),
        ])->save();

        return $this->serializeComment($comment);
    }

    public function comments(int $id, int $page, int $limit): array
    {
        $this->findShortform($id);
        $result = ShortformComment::query()->where('shortform_id', $id)->orderBy('id')->paginate($limit, ['*'], 'page', $page);

        return [
            'items' => $result->getCollection()->map(fn (ShortformComment $comment): array => $this->serializeComment($comment))->all(),
            'pagination' => $this->pagination($result),
        ];
    }

    public function ranking(string $period, int $limit): array
    {
        return Shortform::query()->orderByDesc('like_count')->orderByDesc('view_count')->limit($limit)->get()->map(fn (Shortform $shortform): array => array_merge($this->serializeShortform($shortform), ['period' => $period]))->all();
    }

    public function transcodeStatus(int $id): array
    {
        $shortform = $this->findShortform($id);

        return [
            'status' => $shortform->transcode_status,
            'errorMessage' => null,
            'transcodedAt' => optional($shortform->updated_at)?->toISOString(),
        ];
    }

    public function retry(User $user, int $id): array
    {
        $shortform = $this->findOwnedShortform($user, $id);
        $shortform->forceFill(['transcode_status' => 'QUEUED'])->save();

        return [
            'message' => '트랜스코딩 재시도가 등록되었습니다.',
            'queued' => true,
        ];
    }

    public function delete(User $user, int $id): array
    {
        $shortform = $this->findOwnedShortform($user, $id);
        $shortform->delete();

        return ['message' => '숏폼이 삭제되었습니다.'];
    }

    public function byUser(int $userId, int $page, int $limit): array
    {
        $result = Shortform::query()->where('user_id', $userId)->orderByDesc('id')->paginate($limit, ['*'], 'page', $page);

        return [
            'items' => $result->getCollection()->map(fn (Shortform $shortform): array => $this->serializeShortform($shortform))->all(),
            'pagination' => $this->pagination($result),
        ];
    }

    private function findShortform(int $id): Shortform
    {
        $shortform = Shortform::query()->find($id);
        if (! $shortform) {
            throw new BusinessException('숏폼을 찾을 수 없습니다.', 'SHORTFORM_NOT_FOUND', Response::HTTP_NOT_FOUND);
        }

        return $shortform;
    }

    private function findOwnedShortform(User $user, int $id): Shortform
    {
        $shortform = Shortform::query()->where('user_id', $user->id)->find($id);
        if (! $shortform) {
            throw new BusinessException('숏폼을 찾을 수 없습니다.', 'SHORTFORM_NOT_FOUND', Response::HTTP_NOT_FOUND);
        }

        return $shortform;
    }

    private function serializeShortform(Shortform $shortform): array
    {
        return [
            'id' => $shortform->id,
            'userId' => $shortform->user_id,
            'title' => $shortform->title,
            'videoUrl' => $shortform->video_url,
            'thumbnailUrl' => $shortform->thumbnail_url,
            'viewCount' => $shortform->view_count,
            'likeCount' => $shortform->like_count,
            'commentCount' => $shortform->comment_count,
            'transcodeStatus' => $shortform->transcode_status,
            'createdAt' => optional($shortform->created_at)?->toISOString(),
            'updatedAt' => optional($shortform->updated_at)?->toISOString(),
        ];
    }

    private function serializeComment(ShortformComment $comment): array
    {
        return [
            'id' => $comment->id,
            'shortformId' => $comment->shortform_id,
            'userId' => $comment->user_id,
            'content' => $comment->content,
            'createdAt' => optional($comment->created_at)?->toISOString(),
        ];
    }

    private function pagination($result): array
    {
        return [
            'page' => $result->currentPage(),
            'limit' => $result->perPage(),
            'total' => $result->total(),
            'totalPages' => $result->lastPage(),
        ];
    }
}

File: BackEnd/Java/java-spring-maven-jpa-postgresql/src/main/java/com/pbshop/java/spring/maven/jpa/postgresql/shortform/ShortformService.java
package com.pbshop.java.spring.maven.jpa.postgresql.shortform;

import java.util.List;
import java.util.LinkedHashMap;
import java.util.Map;

import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.pbshop.java.spring.maven.jpa.postgresql.auth.security.AuthenticatedUserPrincipal;
import com.pbshop.java.spring.maven.jpa.postgresql.common.exception.BusinessException;
import com.pbshop.java.spring.maven.jpa.postgresql.common.exception.ErrorCode;
import com.pbshop.java.spring.maven.jpa.postgresql.product.Product;
import com.pbshop.java.spring.maven.jpa.postgresql.product.ProductRepository;
import com.pbshop.java.spring.maven.jpa.postgresql.shortform.dto.ShortformDtos.CommentRequest;
import com.pbshop.java.spring.maven.jpa.postgresql.shortform.dto.ShortformDtos.CreateShortformRequest;
import com.pbshop.java.spring.maven.jpa.postgresql.user.User;
import com.pbshop.java.spring.maven.jpa.postgresql.user.UserRepository;

@Service
@Transactional
public class ShortformService {
    private final ShortformRepository shortformRepository;
    private final ShortformProductRepository shortformProductRepository;
    private final ShortformLikeRepository shortformLikeRepository;
    private final ShortformCommentRepository shortformCommentRepository;
    private final ProductRepository productRepository;
    private final UserRepository userRepository;

    public ShortformService(ShortformRepository shortformRepository, ShortformProductRepository shortformProductRepository,
            ShortformLikeRepository shortformLikeRepository, ShortformCommentRepository shortformCommentRepository,
            ProductRepository productRepository, UserRepository userRepository) {
        this.shortformRepository = shortformRepository;
        this.shortformProductRepository = shortformProductRepository;
        this.shortformLikeRepository = shortformLikeRepository;
        this.shortformCommentRepository = shortformCommentRepository;
        this.productRepository = productRepository;
        this.userRepository = userRepository;
    }

    public Map<String, Object> create(AuthenticatedUserPrincipal principal, CreateShortformRequest request) {
        User user = requireCurrentUser(principal);
        Shortform shortform = new Shortform();
        shortform.setUser(user);
        shortform.setTitle(request.title());
        shortform.setVideoUrl(request.videoUrl());
        shortform.setThumbnailUrl(request.thumbnailUrl());
        Shortform saved = shortformRepository.save(shortform);
        if (request.productIds() != null) {
            for (Long productId : request.productIds()) {
                Product product = productRepository.findById(productId).orElse(null);
                if (product != null) {
                    ShortformProduct relation = new ShortformProduct();
                    relation.setShortform(saved);
                    relation.setProduct(product);
                    shortformProductRepository.save(relation);
                }
            }
        }
        return toResponse(saved);
    }
    @Transactional(readOnly = true)
    public Map<String, Object> feed() {
        List<Map<String, Object>> items = shortformRepository.findAllByOrderByIdDesc().stream().map(this::toResponse).toList();
        Map<String, Object> response = new LinkedHashMap<>();
        response.put("items", items);
        response.put("nextCursor", null);
        return response;
    }
    @Transactional(readOnly = true)
    public Map<String, Object> show(Long id) {
        Shortform shortform = requireShortform(id);
        shortform.setViewCount(shortform.getViewCount() + 1);
        return toResponse(shortform);
    }
    public Map<String, Object> toggleLike(AuthenticatedUserPrincipal principal, Long id) {
        User user = requireCurrentUser(principal);
        Shortform shortform = requireShortform(id);
        boolean liked;
        var like = shortformLikeRepository.findByShortformIdAndUserId(id, user.getId()).orElse(null);
        if (like != null) {
            shortformLikeRepository.delete(like);
            liked = false;
        } else {
            ShortformLike newLike = new ShortformLike();
            newLike.setShortform(shortform);
            newLike.setUser(user);
            shortformLikeRepository.save(newLike);
            liked = true;
        }
        shortform.setLikeCount((int) shortformLikeRepository.countByShortformId(id));
        return Map.of("liked", liked, "likeCount", shortform.getLikeCount());
    }
    public Map<String, Object> addComment(AuthenticatedUserPrincipal principal, Long id, CommentRequest request) {
        User user = requireCurrentUser(principal);
        Shortform shortform = requireShortform(id);
        ShortformComment comment = new ShortformComment();
        comment.setShortform(shortform);
        comment.setUser(user);
        comment.setContent(request.content());
        ShortformComment saved = shortformCommentRepository.save(comment);
        shortform.setCommentCount((int) shortformCommentRepository.countByShortformId(id));
        return toComment(saved);
    }
    @Transactional(readOnly = true)
    public Map<String, Object> comments(Long id) {
        requireShortform(id);
        List<Map<String, Object>> items = shortformCommentRepository.findByShortformIdOrderByIdAsc(id).stream().map(this::toComment).toList();
        return Map.of("items", items, "pagination", Map.of("page", 1, "limit", items.isEmpty() ? 20 : items.size(), "total", items.size(), "totalPages", items.isEmpty() ? 0 : 1));
    }
    @Transactional(readOnly = true)
    public List<Map<String, Object>> ranking(String period) {
        return shortformRepository.findAllByOrderByIdDesc().stream().map(item -> {
            Map<String, Object> response = new LinkedHashMap<>();
            response.put("id", item.getId());
            response.put("period", period);
            response.put("title", item.getTitle());
            response.put("viewCount", item.getViewCount());
            response.put("likeCount", item.getLikeCount());
            response.put("commentCount", item.getCommentCount());
            response.put("videoUrl", item.getVideoUrl());
            response.put("thumbnailUrl", item.getThumbnailUrl() == null ? "" : item.getThumbnailUrl());
            response.put("userId", item.getUser().getId());
            response.put("transcodeStatus", item.getTranscodeStatus());
            response.put("createdAt", item.getCreatedAt() == null ? null : item.getCreatedAt().toString());
            response.put("updatedAt", item.getUpdatedAt() == null ? null : item.getUpdatedAt().toString());
            return response;
        }).toList();
    }
    @Transactional(readOnly = true)
    public Map<String, Object> transcodeStatus(Long id) {
        Shortform shortform = requireShortform(id);
        Map<String, Object> response = new LinkedHashMap<>();
        response.put("status", shortform.getTranscodeStatus());
        response.put("errorMessage", null);
        response.put("transcodedAt", shortform.getUpdatedAt() == null ? null : shortform.getUpdatedAt().toString());
        return response;
    }
    public Map<String, Object> retry(AuthenticatedUserPrincipal principal, Long id) { Shortform shortform = requireOwnedShortform(principal, id); shortform.setTranscodeStatus("QUEUED"); return Map.of("message", "트랜스코딩 재시도가 등록되었습니다.", "queued", true); }
    public Map<String, Object> delete(AuthenticatedUserPrincipal principal, Long id) { Shortform shortform = requireOwnedShortform(principal, id); shortformProductRepository.deleteByShortformId(id); shortformRepository.delete(shortform); return Map.of("message", "숏폼이 삭제되었습니다."); }
    @Transactional(readOnly = true)
    public Map<String, Object> byUser(Long userId) { List<Map<String, Object>> items = shortformRepository.findByUserIdOrderByIdDesc(userId).stream().map(this::toResponse).toList(); return Map.of("items", items, "pagination", Map.of("page", 1, "limit", items.isEmpty() ? 20 : items.size(), "total", items.size(), "totalPages", items.isEmpty() ? 0 : 1)); }

    private Shortform requireShortform(Long id) { return shortformRepository.findById(id).orElseThrow(() -> new BusinessException(ErrorCode.NOT_FOUND, "숏폼을 찾을 수 없습니다.")); }
    private Shortform requireOwnedShortform(AuthenticatedUserPrincipal principal, Long id) { Shortform shortform = requireShortform(id); User user = requireCurrentUser(principal); if (!shortform.getUser().getId().equals(user.getId())) { throw new BusinessException(ErrorCode.NOT_FOUND, "숏폼을 찾을 수 없습니다."); } return shortform; }
    private User requireCurrentUser(AuthenticatedUserPrincipal principal) { if (principal == null) { throw new BusinessException(ErrorCode.UNAUTHORIZED, "인증이 필요합니다."); } return userRepository.findById(principal.userId()).orElseThrow(() -> new BusinessException(ErrorCode.NOT_FOUND, "사용자를 찾을 수 없습니다.")); }
    private Map<String, Object> toResponse(Shortform shortform) {
        Map<String, Object> response = new LinkedHashMap<>();
        response.put("id", shortform.getId());
        response.put("userId", shortform.getUser().getId());
        response.put("title", shortform.getTitle());
        response.put("videoUrl", shortform.getVideoUrl());
        response.put("thumbnailUrl", shortform.getThumbnailUrl() == null ? "" : shortform.getThumbnailUrl());
        response.put("viewCount", shortform.getViewCount());
        response.put("likeCount", shortform.getLikeCount());
        response.put("commentCount", shortform.getCommentCount());
        response.put("transcodeStatus", shortform.getTranscodeStatus());
        response.put("createdAt", shortform.getCreatedAt() == null ? null : shortform.getCreatedAt().toString());
        response.put("updatedAt", shortform.getUpdatedAt() == null ? null : shortform.getUpdatedAt().toString());
        return response;
    }
    private Map<String, Object> toComment(ShortformComment comment) {
        Map<String, Object> item = new LinkedHashMap<>();
        item.put("id", comment.getId());
        item.put("shortformId", comment.getShortform().getId());
        item.put("userId", comment.getUser().getId());
        item.put("content", comment.getContent());
        item.put("createdAt", comment.getCreatedAt() == null ? null : comment.getCreatedAt().toString());
        return item;
    }
}
